components/server/src/api/scm-service-api.ts

```diff
@@ -0,0 +1,93 @@
+/**
+ * Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { HandlerContext, ServiceImpl } from "@connectrpc/connect";
+import { SCMService as ScmServiceInterface } from "@gitpod/public-api/lib/gitpod/v1/scm_connect";
+import { inject, injectable } from "inversify";
+import { PublicAPIConverter } from "@gitpod/gitpod-protocol/lib/public-api-converter";
+import { ScmService } from "../scm/scm-service";
+import {
+    GetSCMTokenRequest,
+    GetSCMTokenResponse,
+    GuessTokenScopesRequest,
+    GuessTokenScopesResponse,
+    SearchRepositoriesRequest,
+    SearchRepositoriesResponse,
+    ListSuggestedRepositoriesRequest,
+    ListSuggestedRepositoriesResponse,
+} from "@gitpod/public-api/lib/gitpod/v1/scm_pb";
+import { ctxUserId } from "../util/request-context";
+import { ApplicationError, ErrorCodes } from "@gitpod/gitpod-protocol/lib/messaging/error";
+import { validate as uuidValidate } from "uuid";
+import { ProjectsService } from "../projects/projects-service";
+import { WorkspaceService } from "../workspace/workspace-service";
+import { PaginationResponse } from "@gitpod/public-api/lib/gitpod/v1/pagination_pb";
+
+@injectable()
+export class ScmServiceAPI implements ServiceImpl<typeof ScmServiceInterface> {
+    constructor(
+        @inject(PublicAPIConverter) private readonly apiConverter: PublicAPIConverter,
+        @inject(ScmService) private readonly scmService: ScmService,
+        @inject(ProjectsService) private readonly projectService: ProjectsService,
+        @inject(WorkspaceService) private readonly workspaceService: WorkspaceService,
+    ) {}
+
+    async getSCMToken(request: GetSCMTokenRequest, _: HandlerContext): Promise<GetSCMTokenResponse> {
+        const userId = ctxUserId();
+        try {
+            const token = await this.scmService.getToken(userId, request);
+            return new GetSCMTokenResponse({ token: this.apiConverter.toSCMToken(token) });
+        } catch (error) {
+            throw new ApplicationError(ErrorCodes.NOT_FOUND, "Token not found.");
```

Comment:
we swallow an error here, can we validate whether it is actual not found and otherwise propagate as is to get an internal error?

Reply:
I somehow feel that it should be part of ScmService.getToken? i.e. guessTokenScopes should also have such contract under the hood?